### DISCUSSION TITLE
DNS Challenge TXT record for hosting local services with valid certificate

### DISCUSSION THREAD

**Discussion by silycr** [tobychui/zoraxy#310] [Q&A]
I would like to use zoraxy for serving local services with a valid certificate. My existing Caddy service generated certificates automatically using DNS-01 challenge with a a DNS TXT record on through Cloudflare.
I notice that zoraxy allows DNS challenge, which successfully issues a certificate through the ACME tool. When setting up a Proxy, I am asked if I'd like to generate a certificate. This process fails:

`Error: one or more domains had a problem: [paperless.dynagen.solutions] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: NXDOMAIN looking up A for paperless.dynagen.solutions - check that a DNS record exists for this domain; DNS problem: NXDOMAIN looking up AAAA for paperless.dynagen.solutions - check that a DNS record exists for this domain`

No TXT record is added. I have not added and propogated a DNS entry, or opened ports to my server as it is not intended to be publicly accessible.

Any suggestions or something I am missing in order to use this to proxy local only services?

**Comment by tobychui** [comment]
Zoraxy default prompt for "generating a certificate for new proxy rules" use HTTP challenge. It only works when your domain is publicly accessible via your preferred CA. So what you are seeing here is expected behavior. 

If you are using local only services in a trusted environment, you do not need HTTPS. In some edge cases where you need a certificate for WebRTC or special web technology support like webcam access, you can create the proxy rule and manually request a certificate using the ACME tool with DNS challenge mode.

**Comment by silycr** [comment]
For my environment, I would like to use HTTPS. At the least it is nice to remove the browser warnings. The additional encrypt/decrypt processing is not a concern for me either.
Given the AuthEmail and AuthToken are saved for a given domain, is it possible to add the function where a certificate can be generate for subdomains using DNS-01 challenge.

Workflow could be:

Open ACME Tool
- Enter Domain "foo.bar"
- CA = Cloudflare
- Use DNS Challenge
- DNS Cred - AuthEmail + AuthToken
- "Save Domain Credentials" (button next to Get Certificate, as a cert is not necessary yet)

Then... Proxy a subdomain "webserver.foo.bar"
- "Generate Certificate" using DNS-01 challenge with saved credentials

**Reply by tobychui** [reply]
The certificate generation function in create proxy rule is done by calling the "Request Certificate" function developed by the ACME module maintainer @yeungalan. If he is ok with your new proposed design (of adding automatic DNS challenge detection) and provide a function call for me, I can fine tune the automation steps regarding the proxy rule manager. 

@yeungalan What do you think?
